### DISCUSSION TITLE
Zoom Call for Collaborators

### DISCUSSION THREAD

**Discussion by ColtAllen** [ColtAllen/btyd#4] [Announcements]
I will be hosting a Zoom call soon to discuss the development direction of this library with other collaborators. My preference would be 
to do this on a Sunday - morning hours for those in the Americas, evening hours for those in Europe.

If you would like to join and have a day preference or any specific topics to discuss, please let me know here or in the issue discussion below:
https://github.com/CamDavidsonPilon/lifetimes/issues/414

**Comment by ColtAllen** [comment]
Zoom call is scheduled for Sunday, 27-Mar at 10 AM Mountain Standard Time (UTC - 6:00):

Join Zoom Meeting
[https://us02web.zoom.us/j/81938221716](https://www.google.com/url?q=https://us02web.zoom.us/j/81938221716&sa=D&source=calendar&usd=2&usg=AOvVaw1BtE0Vcli1FCgVFNmgQ1IP)

Meeting ID: 819 3822 1716
One tap mobile
+12532158782,,81938221716# US (Tacoma)
+13462487799,,81938221716# US (Houston)

Dial by your location
        +1 253 215 8782 US (Tacoma)
        +1 346 248 7799 US (Houston)
        +1 669 900 6833 US (San Jose)
        +1 301 715 8592 US (Washington DC)
        +1 312 626 6799 US (Chicago)
        +1 929 436 2866 US (New York)
Meeting ID: 819 3822 1716
Find your local number: [https://us02web.zoom.us/u/kCp1rZoUe](https://www.google.com/url?q=https://us02web.zoom.us/u/kCp1rZoUe&sa=D&source=calendar&usd=2&usg=AOvVaw2NqoOaksQKh7SA2WJcMkNK)

**Reply by ColtAllen** [reply]
Thanks everyone for attending the Zoom call today. Here's a summary of what we discussed:

### Identified Library Issues
- Python syntax does not include type hinting.
- Instability with `scipy.hypf21`  when using `pandas` inputs, particularly with `GammaGammaFitter`
- Lack of options for plotting & quantifying uncertainty.
- No standard error estimation of Hessian matrix during inference/optimization.
- RFM aggregations computationally prohibitive with large datasets.
- Plotting functions have extraneous dependencies on other methods in the library, limiting flexibility.
- Difficult to determine if calibration and holdout datasets are equally distributed since they can only be split by time period, and is an incomplete approach to model evaluation.
- MLE convergence not stable:
    - `autograd` dependency deprecated two years ago.
    - Current Log-likelihood formulations can cause optimizers to crash.
    - Current MLE penalizer assumptions ill-suited for parameter estimation.
    - Model assumptions not being tested.

### Development Priorities
1. Update documentation to add the contents of this message, updated contributor's guide, and links to research papers.
2. Coveralls Integration for testing coverage (I'm working on this now.)
3. Merge PR of BetaGeo Time-invariant Covariates model submitted by @meremeev.
4. Add type hinting, and separate utility method dependencies from plotting methods.
5. pymc backend integration into `BaseFitter` class. Current MLE approach will be replaced with 'find_MAP' function in `pymc4`, which is expected to be released Apr/May 2022.
6.  Expand model evaluations with the Gelman-Rubin statistic, posterior predictive checks, and other methods. 
7. As the `pymc4` overhaul is ongoing, be mindful of existing issues that have been identified, like log-likelihood formulations and `scipy.hypf21` contributing to convergence instability. Add bug fixes whenever these problems arise. 
 
### Future Additions
* Support for Hierarchical Bayesian models
* Additional models
* Poetry for model packaging
* Nox for version testing (cannot use with poetry, so either of these two must be resolved)
* Distribute RFM aggregation with dask
* Stan backend integration (this will add considerable overhead to the project. If I don't get a lot of requests and PRs related to Stan I will not be pursuing this)

**Comment by guilhermeprokisch** [comment]
Hi!
I'm want to colaborate!
It's gonna have another meeting?

**Reply by ColtAllen** [reply]
Hey @guilhermeprokisch,

Apologies for my delayed response. I've been familiarizing myself with the `pymc` library, which will serve as the new backend for `lifetimes`. `pymc` also has an outstanding developer community, and they regularly host Zoom calls for questions & discussions. Hopefully you can make it to their next event, which I plan on attending:

https://www.meetup.com/pymc-online-meetup/events/285640500/

I'm nearly done with whiteboarding the new modeling API, and aim to get a PR up by the end of this week. After a brief period for feedback from other contributors, I intend to merge it and publish an alpha release to PyPi. Afterward I'll plan another collaborator Zoom call, hopefully before 20-Jun. Let me know if you have a day/time preference.

**Reply by guilhermeprokisch** [reply]
Hey @ColtAllen 

That's sound a good plan! I will wait then your PR to get the context how the project it will work. I had some basic experience in PyMC then I hope it can be useful in the project contribution. 

I don't have any preferred day in Jun for the meeting, any day works for instance. I will keep track the repo for updates.. cheers!

**Comment by ColtAllen** [comment]
@guilhermeprokisch  the PR for the new modeling backend based on `pymc` is now up!

https://github.com/ColtAllen/lifetimes/pull/24
